corehq/apps/app_manager/models.py

```diff
@@ -994,9 +994,113 @@ def validate(self, value, required=True):
+        # doesn't have media set
+        assert media_attr in ('media_image', 'media_audio')
+
+        media_dict = getattr(self, media_attr)
+        if not media_dict:
+            return None
+        if lang in media_dict:
+            return media_dict[lang]
+        elif not strict:
+            # if the queried lang key doesn't exist,
+            # return the first in the sorted list
+            for lang, item in sorted(media_dict.items()):
+                return item
+
+    @property
+    def default_media_image(self):
+        # For older apps that were migrated
+        return self.icon_by_language('default')
+
+    @property
+    def default_media_audio(self):
+        # For older apps that were migrated
+        return self.audio_by_language('default')
+
+    def icon_by_language(self, lang, strict=False):
+        return self._get_media_by_language('media_image', lang, strict=strict)
+
+    def audio_by_language(self, lang, strict=False):
+        return self._get_media_by_language('media_audio', lang, strict=strict)
+
+    def _set_media(self, media_attr, lang, media_path):
+        """
+            Caller's responsibility to save doc.
+            Currently only called from the view which saves after all Edits
+        """
+        assert media_attr in ('media_image', 'media_audio')
+
+        media_dict = getattr(self, media_attr) or {}
+        media_dict[lang] = media_path or ''
+        setattr(self, media_attr, media_dict)
+
+    def set_icon(self, lang, icon_path):
+        self._set_media('media_image', lang, icon_path)
+
+    def set_audio(self, lang, audio_path):
+        self._set_media('media_audio', lang, audio_path)
+
+    def _all_media_paths(self, media_attr):
+        assert media_attr in ('media_image', 'media_audio')
+        media_dict = getattr(self, media_attr) or {}
+        valid_media_paths = {media for media in media_dict.values() if media}
+        return list(valid_media_paths)
+
+    def all_image_paths(self):
+        return self._all_media_paths('media_image')
+
+    def all_audio_paths(self):
+        return self._all_media_paths('media_audio')
+
+    def icon_app_string(self, lang, for_default=False):
+        """
+            Return lang/app_strings.txt translation for given lang
+            - if a path exists for the lang, returns the path
+              otherwise None
+            - if for_default is True (default/app_strings.txt), returns
+              an invalid path deliberately, so that if media is set
+              for one lang but not other, the media for other lang will
+              not be shown. Mobile prefers an invalid path over empty path
+              [Hacky - see http://manage.dimagi.com/default.asp?176008]
+        """
+
+        if not for_default and self.icon_by_language(lang, strict=True):
+            return self.icon_by_language(lang, strict=True)
+
+        if for_default and any(self.all_image_paths()):
```

Comment:
is it necessary to check `any(self.all_image_paths())`? Won't `self.icon_by_language(lang, strict=False)` just return None if there aren't any?